### PR TITLE
NDP/1b: Allow empty values when calculating primary/introduced values

### LINKED ISSUE #3151: NDP: Subcategory value calculation
When calculating subcategory values, remove the rule that all percentages must be filled.

### PR DESCRIPTION
**Pull request by sorja**
image

image

### CODE REVIEW

File: src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+     introduced_area AS (SELECT c.id,
+                                c.country_iso,
+                                c.year,
+                                sum(((c.class ->> 'area'::text)::numeric) *
+                                    ((c.class ->> 'forestPercent'::text)::numeric) /
+                                     100::numeric * ((c.class ->> 'forestPlantationPercent'::text)::numeric) /
+                                     100::numeric *
+                                     ((c.class ->> 'forestPlantationIntroducedPercent'::text)::numeric) /
+                                     100::numeric)
+                                as plantation_forest_introduced_area
```
Comment (sorja) [inline src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts]: changed

File: src/server/schedulers/index.ts [hunk 2]
```diff
@@ -15,6 +15,8 @@ export const initSchedulers = (): void => {
 }
 
 process.on('SIGTERM', async () => {
+  Logger.debug('[System schedulers] SIGTERM received')
   await Promise.all(Object.values(workers).map((worker) => worker.close()))
   Logger.debug('[System schedulers] all workers closed')
+  process.exit()
```
Comment (minotogna) [inline src/server/schedulers/index.ts:21]: Why is this necessary ?

File: src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+       rv.id
+FROM raw_values_2 rv
+         LEFT JOIN total_land_area tla
+                   ON tla.country_iso::text = rv.country_iso::text AND tla.col_name::text = rv.year::text
+         LEFT JOIN primary_forest pf
+                   ON pf.country_iso::text = rv.country_iso::text AND pf.year::text = rv.year::text
+         LEFT JOIN introduced_area ia
+                   ON ia.country_iso::text = rv.country_iso::text AND ia.year::text = rv.year::text;
+      `
+
```
Comment (minotogna) [inline src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts]: Don't we have only one place to handle odp view ?

File: src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    // eslint-disable-next-line no-await-in-loop
+    const countryISOs = (await AreaController.getCountries({ assessment, cycle }, client)).map((c) => c.countryIso)
+
+    // eslint-disable-next-line no-await-in-loop
+    const tables = {
+      [TableNames.originalDataPointValue]: {},
+    }
+    const getCountriesDataProps = { assessment, cycle, countryISOs, tables, force: true }
+    // eslint-disable-next-line no-await-in-loop
+    await DataRedisRepository.getCountriesData(getCountriesDataProps)
```
Comment (sorja) [inline src/test/migrations/steps/20231018130522-step-odp-view-update-primaryforest-introducedforest.ts]: pass client here

**Review by minotogna — APPROVED** [review]
